circom-prover/src/witness.rs

```diff
@@ -16,32 +19,59 @@ pub enum WitnessFn {
     RustWitness(RustWitnessWtnsFn),
 }
 
-pub fn generate_witness(
-    witness_fn: WitnessFn,
-    inputs: HashMap<String, Vec<String>>,
-) -> JoinHandle<Vec<BigUint>> {
-    std::thread::spawn(move || {
-        let bigint_inputs = inputs
-            .into_iter()
-            .map(|(k, v)| {
-                (
-                    k,
-                    v.into_iter()
-                        .map(|i| BigInt::from_str(&i).unwrap())
-                        .collect(),
-                )
-            })
-            .collect();
+pub fn generate_witness(witness_fn: WitnessFn, input_str: String) -> JoinHandle<Vec<BigUint>> {
+    #[cfg(feature = "rustwitness")]
+    let witness_map = json_to_hashmap(input_str.as_str()).unwrap();
 
+    std::thread::spawn(move || {
         let witness = match witness_fn {
             #[cfg(feature = "witnesscalc")]
-            WitnessFn::WitnessCalc(wit_fn) => wit_fn(bigint_inputs),
+            WitnessFn::WitnessCalc(wit_fn) => {
+                let witness = wit_fn(input_str.as_str()).unwrap();
+                parse_witness_to_bigints(&witness).unwrap()
+            }
             #[cfg(feature = "rustwitness")]
-            WitnessFn::RustWitness(wit_fn) => wit_fn(bigint_inputs),
+            WitnessFn::RustWitness(wit_fn) => {
```

Comment:
Note: We should unify the input type of witness generation to `&str` in the future PR.

Reply:
created https://github.com/zkmopro/mopro/issues/346 to follow up.